### PR TITLE
Check for correct env keys in `Command`

### PR DESCRIPTION
**Pull request by tbu-**
Return an error if the key contains an ASCII equals sign (`=`) at the
non-first position.

### CODE REVIEW

**Comment by frewsxcv** [comment]
Tests?

**Comment by alexcrichton** [comment]
Looks good to me, can you be sure to squash the commits as well?

**Comment by tbu-** [comment]
Squashed and rebased.

**Comment by frewsxcv** [comment]
@bors r-

This fails on Windows

https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1822/job/47x30lo5vajnd8pf

**Comment by alexcrichton** [comment]
ping @tbu-, want to help fix the tests?

File: src/libstd/sys/windows/process.rs [hunk 1]
```diff
@@ -35,12 +35,6 @@ use sys_common::{AsInner, FromInner};
 // Command
 ////////////////////////////////////////////////////////////////////////////////
 
-fn mk_key(s: &OsStr) -> OsString {
-    FromInner::from_inner(sys::os_str::Buf {
-        inner: s.as_inner().inner.to_ascii_uppercase()
```
Comment (tbu-) [inline src/libstd/sys/windows/process.rs]: I have no idea why this is here. It ASCII-uppercases all env keys, for no particular reason as far as I can tell -- It was introduced with the stabilisation of the process module, it wasn't there before (if my git is good enough).

Comment (alexcrichton) [inline src/libstd/sys/windows/process.rs]: Can the existing behavior be preserved? We can discuss the specifics in a further issue if you'd like to make a case for removal.

AFAIK Windows has case insensitive keys, so we canonicalize to uppercase.

**Comment by tbu-** [comment]
Rebased and addressed comment.

**Comment by tbu-** [comment]
Removed that semicolon.

**Comment by tbu-** [comment]
Fixed the variable name (sorry, I don't have a Windows machine to test on right now).

**Comment by alexcrichton** [comment]
@bors: retry

* trying to get https://github.com/rust-lang/rust/pull/39852 landed

**Comment by tbu-** [comment]
Fixed variable name.

**Comment by jmesmon** [comment]
Why is the `=` accepted in the first position? Digging around for documentation in posix and windows hasn't turned anything up.

Also may make sense to document this restriction on `Command::env`.

**Comment by tbu-** [comment]
I don't know how that particular test failed. :) It doesn't do anything with the environment.

**Comment by alexcrichton** [comment]
The error message was added in this PR though, so it looks like this PR did indeed cause that error?

**Comment by tbu-** [comment]
I added some `println!`-debugging so I can see why it fails on Android. Can we test this without a chance of accidently getting merged?

**Comment by alexcrichton** [comment]
If you've got docker locally you can test this out via `./src/ci/docker/run.sh android`, but otherwise instead of sending to bors you can also just add `ALLOW_PR=1` to the travis configuration for this builder. That'll build it on this PR but we should just be sure to remove that before the PR lands to not let it leak into master.

**Comment by alexcrichton** [comment]
@tbu- any update on this PR to debug android?

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with android tests fixed!

**Pull request closed without merging**